trace_decoder/src/lib.rs

```diff
@@ -109,30 +63,402 @@
 //! - if there are withdrawals, one dummy transaction is added at the end, with
 //!   all the withdrawals in it.
 
-#![feature(linked_list_cursors)]
-#![feature(trait_alias)]
-#![feature(iter_array_chunks)]
 #![deny(rustdoc::broken_intra_doc_links)]
-#![deny(missing_debug_implementations)]
-#![deny(missing_docs)]
-
-#[cfg(doc)]
-use {
-    trace_protocol::{BlockTrace, TxnInfo},
-    types::OtherBlockData,
-};
-
-/// Provides debugging tools and a compact representation of state and storage
-/// tries, used in tests.
-pub mod compact;
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+
+/// The broad overview is as follows:
+///
+/// 1. Ethereum nodes emit a bunch of binary [`wire::Instruction`]s, which are
+///    parsed in [`wire`].
+/// 2. They are passed to one of two "frontends", depending on the node
+///    - [`hermez_cdk_erigon`], which contains an [`smt_trie`].
+///    - [`zero_jerigon`], which contains an [`mpt_trie`].
+/// 3. The frontend ([`hermez_cdk_erigon::Frontend`] or
+///    [`zero_jerigon::Frontend`]) is passed to the "backend", which lowers to
+///    [`evm_arithmetization::GenerationInputs`].
+///
+/// Deviations from the specification are signalled with `BUG(spec)` in the
+/// code.
+const _DEVELOPER_DOCS: () = ();
+
 /// Defines the main functions used to generate the IR.
-pub mod decoding;
-mod deserializers;
+mod decoding;
+// TODO(0xaatif): add backend/prod support
 /// Defines functions that processes a [BlockTrace] so that it is easier to turn
 /// the block transactions into IRs.
-pub mod processed_block_trace;
-pub mod trace_protocol;
-/// Defines multiple types used in the other modules.
-pub mod types;
-/// Defines useful functions necessary to the other modules.
-pub mod utils;
+mod processed_block_trace;
+mod type1;
+#[cfg(test)]
+#[allow(dead_code)]
+mod type2;
+mod wire;
+
+use std::collections::HashMap;
+
+use ethereum_types::{Address, U256};
+use evm_arithmetization::proof::{BlockHashes, BlockMetadata};
+use evm_arithmetization::GenerationInputs;
+use keccak_hash::H256;
+use mpt_trie::partial_trie::HashedPartialTrie;
+use serde::{Deserialize, Serialize};
+
+/// Core payload needed to generate a proof for a block. Note that the scheduler
+/// may need to request some additional data from the client along with this in
+/// order to generate a proof.
```

Comment:
I don't know if this is the topic for this PR but I find the old terminology is quite confusing (scheduler, client)

```suggestion
/// Core payload needed to generate proof for a block. Additional data retrievable from the blockchain node (using standard ETH RPC API) may be needed for proof generation.
```

Reply:
I barely know what the domain objects are here - thanks for helping to clear up the docs :)